core/state/statedb.go

```diff
@@ -986,6 +1005,132 @@ func (s *StateDB) clearJournalAndRefund() {
+	)
+	for it.Next(true) {
+		// arbitrary stateSize limit, make it configurable
+		if stateSize+nodeSize > 512*1024*1024 {
+			log.Info("Skip large storage deletion", "address", addr.Hex(), "states", stateSize, "nodes", nodeSize)
+			if metrics.EnabledExpensive {
+				slotDeletionSkip.Inc(1)
+			}
+			return true, nil, nil, nil
+		}
+		if it.Leaf() {
+			slots[common.BytesToHash(it.LeafKey())] = common.CopyBytes(it.LeafBlob())
+			stateSize += common.StorageSize(common.HashLength + len(it.LeafBlob()))
+			continue
+		}
+		if it.Hash() == (common.Hash{}) {
+			continue
+		}
+		nodeSize += common.StorageSize(len(it.Path()) + len(it.NodeBlob()))
+		set.AddNode(it.Path(), trienode.NewWithPrev(common.Hash{}, nil, it.NodeBlob()))
+	}
+	if err := it.Error(); err != nil {
+		return false, nil, nil, err
+	}
+	if metrics.EnabledExpensive {
+		if int64(len(slots)) > slotDeletionMaxCount.Value() {
+			slotDeletionMaxCount.Update(int64(len(slots)))
+		}
+		if int64(stateSize+nodeSize) > slotDeletionMaxSize.Value() {
+			slotDeletionMaxSize.Update(int64(stateSize + nodeSize))
+		}
+		slotDeletionTimer.UpdateSince(start)
+		slotDeletionCount.Mark(int64(len(slots)))
+		slotDeletionSize.Mark(int64(stateSize + nodeSize))
+	}
+	return false, slots, set, nil
+}
+
+// handleDestruction processes all destruction markers and deletes the account
+// and associated storage slots if necessary. There are four possible situations
+// here:
+//
+// (a) the account was not existent and be marked as destructed
+// (b) the account was not existent and be marked as destructed,
+//	   but is resurrected later on in the same block.
+//
+// (c) the account was existent and be marked as destructed
+// (d) the account was existent and be marked as destructed,
+//	   but is resurrected later on in the same block.
+//
+//   - in case (a), nothing needs be deleted, nil to nil transition can be ignored.
+//   - in case (b), nothing needs be deleted, but nil is used as the original
+//     value for newly created account and storages
+//   - in case (c), **original** account along with its storages should be deleted,
+//     with their values be tracked as original value.
+//   - in case (d), **original** account along with its storages should be deleted,
+//     with their values be tracked as original value.
+func (s *StateDB) handleDestruction(nodes *trienode.MergedNodeSet) (map[common.Hash]struct{}, error) {
+	incomplete := make(map[common.Hash]struct{})
+	for addr, prev := range s.stateObjectsDestruct {
+		// The original account was non-existing, and it's marked as destructed
+		// in the scope of block. It can be case (a) or (b).
+		// - for (a), skip it without doing anything.
+		// - for (b), track account's original value as nil. It may overwrite
+		//   the data cached in s.accountsOrigin set by 'updateStateObject'.
+		addrHash := crypto.Keccak256Hash(addr[:])
+		if prev == nil {
+			if _, ok := s.accounts[addrHash]; ok {
+				s.accountsOrigin[addrHash] = nil // case (b)
```

Comment:
I think this assignment is not needed. The `nil` entry for this account should already be contained in `accountsOrigin` at this point.